mlos_core/mlos_core/optimizers/observations.py

```diff
@@ -0,0 +1,361 @@
+        return False
+    elif left is not None and right is not None:
+        if not left.equals(right):
+            return False
+    return True
+
+
+class Observation:
+    """
+    A single observation of a configuration.
+
+    Attributes
+    ----------
+    config : pd.Series
+        The configuration observed.
+    score : pd.Series
+        The score metrics observed.
+    context : Optional[pd.Series]
+        The context in which the configuration was evaluated.
+        Not Yet Implemented.
+    metadata: Optional[pd.Series]
+        The metadata in which the configuration was evaluated
+        Not Yet Implemented.
+    """
+
+    def __init__(
+        self,
+        *,
+        config: pd.Series = pd.Series(),
+        score: pd.Series = pd.Series(),
+        context: Optional[pd.Series] = None,
+        metadata: Optional[pd.Series] = None,
+    ):
+        self.config = config
+        self.score = score
+        self.context = context
+        self.metadata = metadata
+
+    def to_suggestion(self) -> Suggestion:
+        """
+        Converts the observation to a suggestion.
+
+        Returns
+        -------
+        Suggestion
+            The suggestion.
+        """
+        return Suggestion(
+            config=self.config,
+            context=self.context,
+            metadata=self.metadata,
+        )
+
+
+class Observations:
+    """
+    A set of observations of a configuration's score.
+
+    Attributes
+    ----------
+    config : pd.DataFrame
+        Pandas dataframe containing configurations. Column names are the parameter names.
+    score : pd.DataFrame
+        The score metrics observed in a dataframe.
+    context : Optional[pd.Series]
+        The context in which the configuration was evaluated.
+        Not Yet Implemented.
+    metadata: Optional[pd.Series]
+        The metadata in which the configuration was evaluated
+        Not Yet Implemented.
+    """
+
+    def __init__(
+        self,
+        *,
+        config: pd.DataFrame = pd.DataFrame(),
+        score: pd.DataFrame = pd.DataFrame(),
+        context: Optional[pd.DataFrame] = None,
+        metadata: Optional[pd.DataFrame] = None,
+        observations: List[Observation] = [],
+    ):
+
+        if len(observations) > 0:
+            config = pd.concat([obs.config.to_frame().T for obs in observations])
+            score = pd.concat([obs.score.to_frame().T for obs in observations])
+            assert len(config.index) == len(
+                score.index
+            ), "config and score must have the same length"
+
+            if sum([obs.context is None for obs in observations]) == 0:
+                context = pd.concat([obs.context.to_frame().T for obs in observations])  # type: ignore
+            else:
+                context = None
+            if sum([obs.metadata is None for obs in observations]) == 0:
+                metadata = pd.concat([obs.metadata.to_frame().T for obs in observations])  # type: ignore
```

Comment:
```suggestion
                metadata = pd.concat([obs.metadata.to_frame().T for obs in observations])  # type: ignore[specific-rule]
```

Let's not do blanket ignores at the very least, being more specific makes it more clear what's going on here.